### ISSUE TITLE
Visualizer not displaying node

### ISSUE THREAD

**Issue by BradleyA** [ManoMarks/docker-swarm-visualizer#6]
The visualizer does display the visualizer screen with the whale without any nodes in browser (Firefox & Chrome) when entering URL: http://192.168.1.201:5000/

$ uname -a
Linux one 3.16.0-76-generic #98~14.04.1-Ubuntu SMP Fri Jun 24 17:04:54 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

$ docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 21:09:07 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 21:09:07 2016
 OS/Arch:      linux/amd64

Commands entered on host one:
$ docker swarm init
Swarm initialized: current node (0r69ag4og37sjuxqpvrnecbo2) is now a manager.
$ docker run -it -d -p 5000:5000 -e HOST=`hostname -i` -e PORT=5000 -v /var/run/docker.sock:/var/run/docker.sock manomarks/visualizer
c2caeb401a591c18e113ca85cde87461df36705bb6984a36f1f356636aab5255
$ docker ps
CONTAINER ID        IMAGE                  COMMAND             CREATED             STATUS              PORTS                              NAMES
12d847d0bf21        manomarks/visualizer   "npm start"         27 minutes ago      Up 27 minutes       0.0.0.0:5000->5000/tcp, 8080/tcp   grave_lamport

$ docker node ls
ID                           NAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
0r69ag4og37sjuxqpvrnecbo2 *  one   Accepted    Ready   Active        Leader
d0b03wee9kff872tpjbe5kpzs    four  Accepted    Ready   Active 

$ docker logs $(docker ps -lq)
npm info it worked if it ends with ok
npm info using npm@2.14.4
npm info using node@v4.1.2
npm info prestart tutumVisualizer@0.0.1
npm info start tutumVisualizer@0.0.1

> tutumVisualizer@0.0.1 start /app
> node server.js

problem with request: connect ECONNREFUSED /var/run/docker.sock
Error: connect ECONNREFUSED /var/run/docker.sock
    at Object.exports._errnoException (util.js:849:11)
    at exports._exceptionWithHostPort (util.js:872:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1060:14)
problem with request: connect ECONNREFUSED /var/run/docker.sock
Error: connect ECONNREFUSED /var/run/docker.sock
    at Object.exports._errnoException (util.js:849:11)
    at exports._exceptionWithHostPort (util.js:872:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1060:14)
problem with request: connect ECONNREFUSED /var/run/docker.sock
Error: connect ECONNREFUSED /var/run/docker.sock
    at Object.exports._errnoException (util.js:849:11)
    at exports._exceptionWithHostPort (util.js:872:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1060:14)
... continues repeating this error ...

Checked fire wall:
$ sudo ufw status verbose
Status: active
Logging: on (medium)
Default: deny (incoming), allow (outgoing), allow (routed)
New profiles: skip
To                         Action      From
--                         ------      ----
22                         LIMIT IN    192.168.1.0/24
53                         ALLOW IN    192.168.1.0/24
80                         ALLOW IN    192.168.1.0/24
111                        ALLOW IN    192.168.1.0/24
443                        ALLOW IN    192.168.1.0/24
2049                       ALLOW IN    192.168.1.0/24
2375                       ALLOW IN    192.168.1.0/24
2376                       ALLOW IN    192.168.1.0/24
3375                       ALLOW IN    192.168.1.0/24
3376                       ALLOW IN    192.168.1.0/24
4789                       ALLOW IN    192.168.1.0/24
7946                       ALLOW IN    192.168.1.0/24
8300                       ALLOW IN    192.168.1.0/24
8301                       ALLOW IN    192.168.1.0/24
8400                       ALLOW IN    192.168.1.0/24
8500                       ALLOW IN    192.168.1.0/24
8600                       ALLOW IN    192.168.1.0/24
2377                       ALLOW IN    192.168.1.0/24
3000                       ALLOW IN    192.168.1.0/24
5000                       ALLOW IN    192.168.1.0/24

Thank you for your help

**Comment by ManoMarks** [comment]
Looks like your firewall is preventing access to the Docker Unix socket.

**Comment by BradleyA** [comment]
Turning off the firewall on all systems did not correct the incident.
The error messages in the container log do not appear until I enter  URL: http://192.168.1.201:5000/ in the browser and stop when I close the browser and start again when entering the URL: http://192.168.1.201:5000/  in the browser.

**Comment by BradleyA** [comment]
You are correct that the observed symptom of the visualizer not displaying any nodes in browser is the same as #3 but #3 states no errors were found in the log as this one does.  That is why I opened it to post the errors I am viewing in the log.  It would be more helpful if you could suggest additional debugging method you would recommend, thank you.

**Comment by ManoMarks** [comment]
Sorry I'm on vacation right now and unable to test things out

**Comment by ManoMarks** [comment]
I'll be back next week and in front of a computer

**Comment by ManoMarks** [comment]
OK so I'm back. Is it possible your daemon is running on a different socket? It can be configured to run differently. Other things to try: Did you build the visualizer image yourself? Or use the hub version? If you used the hub version, try pulling again I just pushed a new image that works for me.

**Comment by BradleyA** [comment]
Repeated test on another system named three, same incident
Download docker 1.12.0-rc4
Restarted docker
$ sudo service docker restart
Downloaded latest manomarks/visualizer  (did not build my own image)
$ docker images | grep mano
manomarks/visualizer                     latest              7f2899d7f460        23 hours ago        373.5 MB

Output of dockerd
$ ps -ef | grep dockerd
root      4954     1  1 12:23 ?        00:00:31 /usr/bin/dockerd --graph=/usr/local/docker --dns 192.168.1.202 --dns 8.8.8.8 --dns 8.8.4.4 --log-level warn --tlsverify --tlscacert=/etc/docker/certs.d/daemon/ca.pem --tlscert=/etc/docker/certs.d/daemon/three.cptx86.com-cert.pem --tlskey=/etc/docker/certs.d/daemon/three.cptx86.com-priv-key.pem -H=three.cptx86.com:2376 --raw-logs

$ docker logs $(docker ps -lq)
npm info it worked if it ends with ok
npm info using npm@2.14.4
npm info using node@v4.1.2
npm info prestart swarmVisualizer@0.0.1
npm info start swarmVisualizer@0.0.1

> swarmVisualizer@0.0.1 start /app
> node server.js

problem with request: connect ECONNREFUSED /var/run/docker.sock
Error: connect ECONNREFUSED /var/run/docker.sock
    at Object.exports._errnoException (util.js:849:11)
    at exports._exceptionWithHostPort (util.js:872:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1060:14)
problem with request: connect ECONNREFUSED /var/run/docker.sock
Error: connect ECONNREFUSED /var/run/docker.sock
    at Object.exports._errnoException (util.js:849:11)
    at exports._exceptionWithHostPort (util.js:872:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1060:14)
. . .

**Comment by ManoMarks** [comment]
What showed up in the Chrome console logs?

**Comment by ManoMarks** [comment]
Never mind, it's a socket error, happening at initial connection to the docker socket.

**Comment by ManoMarks** [comment]
What's the output of `netstat -a` ? Do you see something like 
`unix  2      [ ACC ]     STREAM     LISTENING      19954 /var/run/docker.sock`
?

**Comment by ManoMarks** [comment]
Also are you in the docker group? Check out the [Docker Docs](https://docs.docker.com/engine/installation/linux/ubuntulinux/) for more info

**Comment by BradleyA** [comment]
$ netstat -a | grep docker
unix  2      [ ACC ]     STREAM     LISTENING     2620     /var/lib/docker/network/files/54191b75bbb522945a8a4db733be3b61fb2fbd15dc5bc67ea6e1eb7e7c91979b.sock
unix  2      [ ACC ]     STREAM     LISTENING     26883    /usr/local/docker/swarm/control.sock
unix  2      [ ACC ]     STREAM     LISTENING     2427     /var/run/docker/libcontainerd/docker-containerd.sock
unix  3      [ ]         STREAM     CONNECTED     21105    /usr/local/docker/swarm/control.sock
unix  3      [ ]         STREAM     CONNECTED     1566     /var/run/docker/libcontainerd/docker-containerd.sock

$ groups
uthree adm cdrom sudo dip plugdev lpadmin sambashare docker

**Comment by ManoMarks** [comment]
What happens if you run `curl --unix-socket /var/run/docker.sock http:/containers/json`

**Comment by ManoMarks** [comment]
Also you can check out the docks on the [Docker Remote API](https://docs.docker.com/engine/reference/api/docker_remote_api/)

**Comment by BradleyA** [comment]
$ curl --unix-socket /var/run/docker.sock http:/containers/json
curl: (7) Couldn't connect to server

**Comment by ManoMarks** [comment]
OK so I think Docker is running on a different socket for you. Figure out what that is and substitute that for /var/run/docker.sock in the code should make it work

**Comment by BradleyA** [comment]
How would I determine what socket docker is running on ?

**Comment by ManoMarks** [comment]
I haven't been able to find a way. But, when you start Docker you can specify the path: `https://docs.docker.com/engine/reference/commandline/dockerd/`

**Comment by BradleyA** [comment]
Do you think the incident is because I am using TLS ? (-H=three.cptx86.com:2376)
Here is the option I an using with the daemon:
$ ps -ef | grep dockerd
root 4954 1 1 12:23 ? 00:00:31 /usr/bin/dockerd --graph=/usr/local/docker --dns 192.168.1.202 --dns 8.8.8.8 --dns 8.8.4.4 --log-level warn --tlsverify --tlscacert=/etc/docker/certs.d/daemon/ca.pem --tlscert=/etc/docker/certs.d/daemon/three.cptx86.com-cert.pem --tlskey=/etc/docker/certs.d/daemon/three.cptx86.com-priv-key.pem -H=three.cptx86.com:2376 --raw-logs

**Comment by ManoMarks** [comment]
Yeah that could be it. I'm not as experienced in using that. Two things to try

1. use curl as described above to the default unix socket, but with certificates
2. try curling three.cptx86.com instead

**Comment by BradleyA** [comment]
1.
curl --cert ~/.docker/cert.pem  --key ~/.docker/key.pem  --cacert ~/.docker/ca.pem --unix-socket /var/run/docker.sock https://`hostname -f`:2376/containers/json
curl: (7) Couldn't connect to server
2.
curl --cert ~/.docker/cert.pem  --key ~/.docker/key.pem  --cacert ~/.docker/ca.pem  https://`hostname -f`:2376/containers/json [{"Id":"d5c03ceaf97183260d021ab21a274f488b72dacba1943b444148bfec17015f37","Names":["/tomcat"],"Image":"tomcat","ImageID":"sha256:25e98610c7d0ce3c3cf05a6d8b3e932135d86720f1d46e73a49a8256eca27892","Command":"cata . . .

**Comment by ManoMarks** [comment]
Ah good. If you curl it without the certificates, then you have the path in your second example. You would just need to change the path in line 39 of server.js. However, if you need the certificates there's probably a bit more work. I'm using the  [ws NPM module](https://github.com/websockets/ws) for websocket connection which does look like it has some certificate support. It looks like I should have environment variable customization for custom daemon routes and certificates. Not sure when I'll get to that but I'll give it a shot. Is it OK to close this issue? I'll start another one to add those features and cc you.

**Comment by BradleyA** [comment]
Thank you for your help.  All my systems are using TLS communications for Docker CLI and daemon.
Closing incident as current release does not supporting TLS.
I have found another method to receive the information I needed using the Linux watch command, but it is tied to a task, see example below:
docker service create --replicas 3 --name redis --update-delay 10s --update-parallelism 1 redis:3.0.6
docker service scale redis=20
docker service update --image redis:3.0.7 redis
watch -d -n 4 docker service tasks redis
